src/ott/neural/methods/expectile_neural_dual.py

```diff
@@ -0,0 +1,526 @@
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Iterator,
+    List,
+    Literal,
+    Optional,
+    Sequence,
+    Tuple,
+    Union,
+)
+
+import jax
+import jax.numpy as jnp
+from jax.lax import stop_gradient
+
+import optax
+from flax import linen as nn
+from flax.core.frozen_dict import FrozenDict
+
+from ott import utils
+from ott.geometry import costs
+from ott.neural.networks.potentials import (
+    BasePotential,
+    PotentialGradientFn_t,
+    PotentialTrainState,
+    PotentialValueFn_t,
+)
+from ott.problems.linear.potentials import DualPotentials
+
+__all__ = [
+    "ENOTPotentials", "PotentialModelWrapper", "MLP", "ExpectileNeuralDual"
+]
+
+Train_t = Dict[Literal["train_logs", "valid_logs"], Dict[str, List[float]]]
+Callback_t = Callable[[int, DualPotentials], None]
+
+
+@jax.tree_util.register_pytree_node_class
+class ENOTPotentials(DualPotentials):
+  """The dual potentials of ENOT method.
+
+  Args:
+    grad_f: Gradient of the first dual potential function.
+    g: The second dual potential function.
+    cost_fn: The cost function used to solve the OT problem.
+    is_bidirectional: Whether the duals are trained for bidirectional
+      transport mapping.
+    corr: Whether the duals solve the problem in correlation form.
+  """
+
+  def __init__(
+      self, grad_f: PotentialGradientFn_t, g: PotentialValueFn_t,
+      cost_fn: costs.CostFn, is_bidirectional: bool, corr: bool
+  ):
+
+    self.__grad_f = grad_f
+    self.is_bidirectional = is_bidirectional
+
+    if is_bidirectional and not corr:
+      transport = lambda x: cost_fn.twist_operator(x, grad_f(x), False)
+    else:
+      transport = lambda x: grad_f(x)
+
+    conjugate_cost = jnp.dot if corr else cost_fn
+
+    def g_cost_conjugate(x: jnp.ndarray) -> jnp.ndarray:
+      y_hat = jax.lax.stop_gradient(transport(x))
+      return -g(y_hat) + conjugate_cost(x, y_hat)
+
+    super().__init__(g_cost_conjugate, g, cost_fn=cost_fn, corr=corr)
+
+  @property
+  def _grad_f(self) -> Callable[[jnp.ndarray], jnp.ndarray]:
+    return jax.vmap(self.__grad_f)
+
+  def transport(self, vec: jnp.ndarray, forward: bool = True) -> jnp.ndarray:
+    """Transport from source to the target distribution or vice-versa."""
+    if self.is_bidirectional:
+      return super().transport(vec, forward)
+
+    vec = jnp.atleast_2d(vec)
+    assert (forward
+            is True), ("Only forward mapping (source -> target) is supported.")
+    return self._grad_f(vec)
```

Comment:
No, `T(x) = grad(f)(x)` in case `self.is_bidirectional=False`  and independent on `corr`.   If `self.is_bidirectional=False`  we can use any function for the transport mapping T(x), including `x - grad(f)(x)`  (since we don't express it through the potential f). But the last transform  ``x - grad(f)(x)`  may be unexpected by user and `grad(f)(x)` seems to be more straightforward. I have also updated it in the new version of paper.

Reply:
Thank you for the explanation, makes perfect sense!